redis/lookup.go

```
@@ -0,0 +1,177 @@
+		err = look.handleDNSError(err, "SRV", srv)
+		if err != nil {
+			tempErrs = append(tempErrs, err)
+			// Skip to the next SRV lookup.
+			continue
+		}
+
+		for _, target := range targets {
+			host := strings.TrimRight(target.Target, ".")
+			if look.dnsAuthority != "" {
+				// Lookup A/AAAA records for the SRV target using the custom DNS
+				// authority.
+				hostAddrs, err := resolver.LookupHost(ctx, host)
+				err = look.handleDNSError(err, "A/AAAA", srv)
+				if err != nil {
+					tempErrs = append(tempErrs, err)
+					// Skip to the next A/AAAA lookup.
+					continue
+				}
+				if len(hostAddrs) == 0 {
+					// Skip to the next A/AAAA lookup.
+					continue
+				}
+				// Use the first resolved IP address.
+				host = hostAddrs[0]
+			}
+			addr := fmt.Sprintf("%s:%d", host, target.Port)
+			newAddrs[addr] = addr
+		}
+	}
+	// Only return an error if all lookups failed.
+	if len(tempErrs) > 0 && len(newAddrs) == 0 {
+		return nil, errors.Join(tempErrs...)
+	}
+	return newAddrs, nil
+}
+
+// shardsPeriodically periodically performs SRV lookups for the given service
+// name and updates the ring shards accordingly.
+func (look *Lookup) shardsPeriodically(ctx context.Context, frequency time.Duration) {
+	ticker := time.NewTicker(frequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			timeoutCtx, cancel := context.WithTimeout(ctx, frequency)
+			newAddrs, err := look.Shards(timeoutCtx)
+			cancel()
+			if err != nil {
+				look.logger.Warningf("resolving redis shards for %+v, temporary errors occurred: %s", look.srvLookups, err)
+				continue
+			}
+			if len(newAddrs) == 0 {
+				look.logger.Errf("0 redis shards were resolved for %+v", look.srvLookups)
+				continue
+			}
+			look.ring.SetAddrs(newAddrs)
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Start begins periodic SRV lookups and updates the ring shards accordingly.
+func (look *Lookup) Start(ctx context.Context) {
```

Why does this panic instead of returning an error? It feels like this was maybe left over from a version where callers were expected to call `go lookup.Start()`, and therefore not be able to extract an error from it? Given that this does one synchronous lookup before kicking off the goroutine, I think it would be idiomatic for failures during that initial lookup to return errors.